python/ray/data/dataset.py

```diff
@@ -2305,15 +2305,23 @@ def take_batch(
         """
         batch_format = _apply_strict_mode_batch_format(batch_format)
         try:
+            from ray.data.dataset_pipeline import DatasetPipeline
+
+            is_not_pipeline = not isinstance(self, DatasetPipeline)
+            limited_ds = self.limit(batch_size) if is_not_pipeline else self
             res = next(
                 iter(
-                    self.iter_batches(
+                    limited_ds.iter_batches(
                         batch_size=batch_size,
                         prefetch_batches=0,
                         batch_format=batch_format,
                     )
                 )
             )
+            if is_not_pipeline:
+                # Save the computed stats and blocks to the original dataset.
+                self._plan._snapshot_stats = limited_ds._plan.stats()
+                self._plan._snapshot_blocks = limited_ds._plan._snapshot_blocks
```

Comment:
Can put the new code outside of the `try-except` block? It'd be clearer which line of code could raise the exception.